### PR TITLE
Update ci

### PR DESCRIPTION
**Pull request by jakkdl**
* bump pre-commit checkers
* use actions/checkout@v4
* use actions/setup-python@v5
* New Black style guide
* fix mypy errors
  * `if ...:` is now treated as `if True:` and it therefore thought tons of the `eval_files` contained unreachable statements. Replaced it with an arbitrary function call.
* fix pyright errors
  * ignored error about `Statement` being unhashable, [which it technically should be](https://docs.python.org/3/reference/datamodel.html#object.__hash__), although I guess we're inheriting a `__hash__` from `NamedTuple` and it currently works :tm:. Need to change a bunch of equality checks in the code to make it less illegal, so leaving that for now.
  * https://github.com/Instagram/LibCST/issues/1107
* ruff
  * bug in TCH002 fixed: https://github.com/astral-sh/ruff/issues/4667
  * 2.0 has new recommended config setup
* remove flake8 from pre-commit (it was only kept for curiosity, and started giving new errors)
* add py312 in CI
* pyupgrade broke a test file and gave invalid code, so that file is now excluded
  * https://github.com/asottile/pyupgrade/issues/932
* TRIO910 "non-empty async function must contain checkpoint" did not properly detect
```python
async def foo(): ...
```
as an empty function, since it was matching against the indented body. This is now the black style, which is why I found it. So fixed that, and updated test file to have both.
There's only a few more instances of `IndentedBlock` in the project, and none seem problematic, but it's possible this could be a problem in other places.
* adding 3.13 to CI is blocked by https://github.com/Zac-HD/hypothesmith/issues/36

I can split this across multiple PRs, but idk if it's warranted. Now that we moved and can set up pre-commit it should be more incremental going forward.

### CODE REVIEW

**Comment by jakkdl** [comment]
@thejcannon @aneeshusa

File: flake8_trio/visitors/helpers.py [hunk 1]
```diff
@@ -29,7 +29,9 @@
 
     T = TypeVar("T", bound=Flake8TrioVisitor)
     T_CST = TypeVar("T_CST", bound=Flake8TrioVisitor_cst)
-    T_EITHER = TypeVar("T_EITHER", bound=Flake8TrioVisitor | Flake8TrioVisitor_cst)
+    T_EITHER = TypeVar(
+        "T_EITHER", bound=Union[Flake8TrioVisitor, Flake8TrioVisitor_cst]
+    )
```
Comment (thejcannon) [inline flake8_trio/visitors/helpers.py:34]: Why this change?

File: tests/autofix_files/trio911.py [hunk 2]
```diff
@@ -13,8 +13,11 @@ async def foo() -> Any:
     await foo()
 
 
-def bar(*args) -> Any:
-    ...
+def bar(*args) -> Any: ...
+
+
+# mypy now treats `if ...` as `if True`, so we have another arbitrary function instead
```
Comment (thejcannon) [inline tests/autofix_files/trio911.py:19]: I mean...

```pycon
>>> bool(...)
True
```

😂

File: tests/autofix_files/trio91x_autofix.py [hunk 3]
```diff
@@ -39,10 +38,10 @@ async def foo_yield():  # TRIO911: 0, "exit", Statement("yield", lineno+2)
 
 
 async def foo_if():
-    if ...:
+    if foo():
```
Comment (thejcannon) [inline tests/autofix_files/trio91x_autofix.py:41]: since `foo` is async, should we `if bar()` in these places?

File: tests/eval_files/trio102.py [hunk 4]
```diff
@@ -120,7 +120,9 @@ async def foo():
     try:
         pass
     finally:
-        async for i in trio.bypasslinters:  # error: 8, Statement("try/finally", lineno-3)
+        async for (  # error: 8, Statement("try/finally", lineno-3)
+            i
+        ) in trio.bypasslinters:
```
Comment (thejcannon) [inline tests/eval_files/trio102.py:125]: What spurred this?

File: tests/eval_files/trio91x_autofix.py [hunk 5]
```diff
@@ -34,9 +33,9 @@ async def foo_yield():  # TRIO911: 0, "exit", Statement("yield", lineno+2)
 
 
 async def foo_if():
-    if ...:
+    if foo():
```
Comment (thejcannon) [inline tests/eval_files/trio91x_autofix.py:36]: Same comment about `async` `foo`

**Review by thejcannon — APPROVED** [review]
Mostly little things

**Review by Zac-HD — APPROVED** [review]
Thanks John!

File: flake8_trio/visitors/helpers.py (hunk 1 above)
Comment (jakkdl) [reply flake8_trio/visitors/helpers.py:34]: I didn't previously specify `python_version` in `pyproject.toml` for mypy, and now that I did it didn't like the old version which wouldn't work at runtime - and mypy doesn't seem to detect that we're inside an `if TYPE_CHECKING` block.

File: flake8_trio/visitors/flake8triovisitor.py [hunk 6]
```diff
@@ -23,7 +23,9 @@
 
 class Flake8TrioVisitor(ast.NodeVisitor, ABC):
     # abstract attribute by not providing a value
-    error_codes: dict[str, str]  # pyright: ignore[reportUninitializedInstanceVariable]
+    error_codes: ClassVar[
+        dict[str, str]
+    ]  # pyright: ignore[reportUninitializedInstanceVariable]
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply flake8_trio/visitors/flake8triovisitor.py:28]: It is not! And turns out `reportUnnecessaryTypeIgnoreComment` doesn't default to true with `strict = true`, so time to turn that on.
I did change this one hoping it would flow through and silence RUF012 ("Mutable class attribute should be annotated with `typing.ClassVar`") to the classes inheriting from it - but when that didn't happen I silenced the error rather than change it in the 25 other places. But maybe I'll bother doing the big search&replace

File: tests/autofix_files/trio91x_autofix.py (hunk 3 above)
Comment (jakkdl) [reply tests/autofix_files/trio91x_autofix.py:41]: huh, I'd expect that to trigger [mypy's unused-coroutine](https://mypy.readthedocs.io/en/latest/error_code_list.html#check-that-coroutine-return-value-is-used-unused-coroutine) but I guess the return value is technically used.
Reported it at https://github.com/python/mypy/issues/16921#issue-2138489669

For the test file it of course doesn't really matter, unless mypy starts treating `if foo():` as `if True:` :grin:, but yeah no reason not to.

File: tests/eval_files/trio102.py (hunk 4 above)
Comment (jakkdl) [reply tests/eval_files/trio102.py:125]: New version of Black now considers the line too long (90>88 after all) and splits it up, after which I move the comment back to line 123.
Not super pretty, maybe worth reporting upstream.

File: flake8_trio/visitors/flake8triovisitor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakkdl) [reply flake8_trio/visitors/flake8triovisitor.py:28]: oh hm, the reason it wasn't enabled is because there's no separate setting for `pyright: ignore` vs `type: ignore` ([and unlikely to be added](https://github.com/python/mypy/issues/12358)), and [mypy doesn't have `mypy: ignore`](https://github.com/python/mypy/issues/12358)

File: flake8_trio/visitors/helpers.py (hunk 1 above)
Comment (jakkdl) [reply flake8_trio/visitors/helpers.py:34]: although I never remember which tools want `python_version` to be set to max-version-you-support vs min-version-you-support, so maybe should set it to 3.12 instead of 3.9

Comment (jakkdl) [reply flake8_trio/visitors/helpers.py:34]: mypy wants you to run separate runs of all of them in parallel x)
https://github.com/python/mypy/issues/12286

File: flake8_trio/visitors/flake8triovisitor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakkdl) [reply flake8_trio/visitors/flake8triovisitor.py]: Could perhaps enable mypy's `type: ignore`, but that'd involve adding a bunch of deps to its pre-commit env and/or go through a bunch of them and replace with `pyright: ignore` - and I don't think that's worth bothering with currently.

File: tests/eval_files/trio91x_autofix.py (hunk 5 above)
Comment (jakkdl) [reply tests/eval_files/trio91x_autofix.py:36]: yeah these are the same file, `eval_files` are the original files, and then the ones in in `autofix_files` are generated (and diff'd against in tests), so these two have to match.

File: flake8_trio/visitors/flake8triovisitor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zac-HD) [reply flake8_trio/visitors/flake8triovisitor.py]: If we annotate as `Mapping[str, str]` so that it's logically immutable maybe our typecheckers will be happier?

File: tox.ini [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
 
 # create a default testenv, whose behaviour will depend on the name it's called with.
 # for CI you can call with `-e flake8_5,flake8_6` and let the CI handle python version
 [testenv]
 description = Runs pytest, optionally with posargs
 deps =
-    flake8_6: flake8>=6.0
-    flake8_5: flake8>=5.0,<6.0
+    flake8_7: flake8>=7.0
+    flake8_6: flake8>=6.0, <7.0
```
Comment (Zac-HD) [inline tox.ini:12]: Unclear from this whether or not we have flake8_5 and _7 environments - the deps list has changed, but not the envlist (or comment).  

Once you confirm that this is working as intended, including in CI, I'll merge 🙂
